src/trio/_core/_ki.py

```diff
@@ -83,18 +77,108 @@
 # for any Python program that's written to catch and ignore
 # KeyboardInterrupt.)
 
-# We use this special string as a unique key into the frame locals dictionary.
-# The @ ensures it is not a valid identifier and can't clash with any possible
-# real local name. See: https://github.com/python-trio/trio/issues/469
-LOCALS_KEY_KI_PROTECTION_ENABLED: Final = "@TRIO_KI_PROTECTION_ENABLED"
+_T = TypeVar("_T")
+
+if TYPE_CHECKING or sys.version_info >= (3, 9):
+    _Ref = weakref.ref[_T]
+else:
+
+    class _Ref(Generic[_T], weakref.ref):
+        __slots__ = ()
+
+
+class _IdRef(_Ref[_T]):
+    __slots__ = ("_hash",)
+    _hash: int
+
+    def __new__(cls, ob: _T, callback: Callable[[Self], Any] | None = None, /) -> Self:
+        self: Self = weakref.ref.__new__(cls, ob, callback)
+        self._hash = object.__hash__(ob)
+        return self
+
+    def __eq__(self, other: object) -> bool:
+        if self is other:
+            return True
+
+        if not isinstance(other, _IdRef):
+            return NotImplemented
+
+        my_obj = None
+        other_obj: Any = None
+        try:
+            my_obj = self()
+            other_obj = other()
+            return my_obj is not None and other_obj is not None and my_obj is other_obj
```

Comment:
Technically `other_obj is not None` is not needed here; if my_obj is not None, then `my_obj is other_obj` implies that other_obj is also not None